fix: Report sidebar must consider Permission Query

marination opened the pull request:
**Issue:**
<img width="1331" alt="Screenshot 2023-01-13 at 5 11 43 PM" src="https://user-images.githubusercontent.com/25857446/212312500-d72f9438-1516-44e2-9d09-fa5f3508bdcf.png">

- Add a Permission Query (Server Script) to the **Report** Doctype (Eg. each user must only view their own custom reports)
- **test task** is a custom report created by another user
- Reports in Report View sidebar does not respect it
- In `boot.py` -> `get_user_pages_or_reports` is used to set all permissible reports in the cache, which is then used to render the Report view sidebar and only takes roles into consideration
- **Expectation:** Reports visible in the sidebar must be a subset of the reports visible to the user in the **Report List View**. Permission query must apply every major area where Reports are fetched.

**Fix:**
<img width="1308" alt="Screenshot 2023-01-13 at 5 12 40 PM" src="https://user-images.githubusercontent.com/25857446/212312667-9e65e783-a686-477c-875a-26973110095e.png">

- While caching permissible reports on boot, filter out reports blocked by Permission Query in `get_user_pages_or_reports`
- Sidebar report selector uses boot cache to get allowed reports, which now respects Permission Query
- Convert qb query to str > append permission query > execute. (Couldn't find any other way to turn str to a pypika query and Permission Query is stored as a string in Server Script)

> **To the reviewer:** Shouldn't permission query be skipped for an Admin user? They should ideally view everything right ?If yes, a fix will have to be made in `DatabaseQuery.execute()` as well as the current PR

marination committed:
fix: Report sidebar must consider Permission Query

- On boot cache permissible reports, filter out reports blocked by Permission Query
- Sidebar report selector uses boot cache to get allowed reports, which now respects Permission Query
- Convert qb query to str and append permission query and then execute

Co-authored-by: Gavin D'souza <gavin18d@gmail.com>

marination committed:
fix: Make `run_with_permission_query` private (not a general util)

- Also switch query and doctype parameter positions in `_run_with_permission_query`

gavindsouza commented on frappe/boot.py:
```diff
@@ -248,6 +254,21 @@ def get_user_pages_or_reports(parent, cache=False):
 	return has_role
 
 
+def _run_with_permission_query(query: "MySQLQueryBuilder", doctype: str) -> list[dict]:
```
```suggestion
def _run_with_permission_query(query: "Query", doctype: str) -> list[dict]:
```

Maybe Query from `frappe.database.utils` is better here?

gavindsouza commented on frappe/boot.py:
```diff
@@ -248,6 +254,21 @@ def get_user_pages_or_reports(parent, cache=False):
+	Adds Permission Query (Server Script) conditions and runs/executes modified query
+	Note: Works only if 'WHERE' is the last clause in the query
+	"""
+	db_query = DatabaseQuery(doctype, frappe.session.user)
+	permission_query = db_query.get_permission_query_conditions()
+
+	query = query.get_sql()
+	if permission_query:
+		query = query + " AND " + permission_query
```
```suggestion
	if permission_query := DatabaseQuery(doctype, frappe.session.user).get_permission_query_conditions():
		query = f"{query} AND {permission_query}"

```

marination replied:
I'll make `permission_query` declaration a one-liner, but the walrus bit is hard to read personally.

marination committed:
refactor: Use `Query` instead of `MySQLQueryBuilder`

- Use `Query` from database/utils which will consider postgres query type as well
- Reduce LOC where unnecessary

Co-authored-by: Gavin D'souza <gavin18d@gmail.com>

marination committed:
Merge branch 'develop' into report-sidebar-respect-perm-query

marination committed:
fix: Remove unnecessary `get_sql`

marination committed:
test: Test if permission query via server script is applied on cached allowed reports

marination committed:
chore: Add comments to avoid incompatible queries with `_run_with_permission_query`

gavindsouza committed:
chore: Remove errprint triggered by passing Query object to db.sql

gavindsouza committed:
Merge branch 'develop' into report-sidebar-respect-perm-query

gavindsouza commented on frappe/tests/test_boot.py:
```diff
@@ -26,3 +26,51 @@ def test_get_unseen_notes(self):
+			# Test user must not see admin user's report
+			self.assertNotIn("Test Admin Report", allowed_reports)
+			self.assertIn("Test User Report", allowed_reports)
+		finally:
+			frappe.db.rollback()
+			frappe.set_user("Administrator")
```
* This `try-finally` is essentially a contextmanager/decorator - I think we can do without the long try block. 
* Getting a PermissionError while running this for test user creating Report, add ignore_permissions?

marination committed:
fix(test): Remove try-finally & ignore perms on test user's report insertion

marination committed:
Merge branch 'develop' into report-sidebar-respect-perm-query

ankush committed:
fix: Only apply perm query to non-admin users
